### PR TITLE
Add the possibility to style Table rows by providing a 'style' attribute

### PR DESCRIPTION
**Pull request by Augustin82**

### CODE REVIEW

File: src/DataTable/Table.js [hunk 1]
```diff
@@ -65,7 +65,7 @@ class Table extends React.Component {
                 </thead>
                 <tbody>
                     {realRows.map((row, idx) =>
-                        <tr key={row[rowKeyColumn] || row.key || idx} className={row.className}>
+                        <tr key={row[rowKeyColumn] || row.key || idx} className={row.className} style={row.style}>
```
Comment (tleunen) [inline src/DataTable/Table.js]: Hmm...

I know we're using that for `row.className` but I wonder if with `style` it could produce errors. Because if the rows data have `style` as a key, and that style is a string for example, React will not be happy.

I wonder if we could find a better way to allow the row styling (even for the className to be honest).

Comment (Augustin82) [inline src/DataTable/Table.js]: So do I... ;)

However, in the meantime, since there is support for className, I believe it only makes sense to have support for style, even if it means handling the entire thing differently later on.

What checks on the nature of 'style' should be performed to make it safe?
What checks on 'className' ?
Also, maybe 'style' should be renamed to something less likely to collide, like 'rowStyle' ?

**Comment by Augustin82** [comment]
@tleunen any update on how I should proceed? Do I need to write more tests, to fix the handling of edge cases, to argue my case some more?

Please let me know and I'll get to work. Cheers!

**Comment by tleunen** [comment]
Sorry @Augustin82, I've been busy with some other things recently and couldn't spend much time on react-mdl :s 

So after a quick investigation (to remind me how the Table works), `className` is mostly used internally to set the row as "selected". While I believe it's ok, I also think it could conflict with the row data, especially if the row data already has a `className` or `style` property.
I believe for v1, we should be fine, but for v2, I'd prefer having a special prop, maybe something like `__mdlRowProp` or something very unlikely to be used as a data prop to avoid any conflict.

Thoughts?

**Comment by Augustin82** [comment]
@tleunen no worries, I was just pinging you. Thanks for taking the time to answer.

I agree with what you said on the need for v2 to have a special prop that would help limiting conflicts (along with explicitly defining it in the documentation).

Would it be OK, for the current version, to release support for "proper" React styling, still? I could update my PR to use a less obvious name than style (such as `__mdlRowClassName` and `__mdlRowStyle`, or something like you suggested).

The same thing could be extended to individual cells if deemed necessary; again, I could update my PR to support this.

Thanks!

**Comment by tleunen** [comment]
On a cell basis, it could be a bit more complicated to do so. To allow that, each cell data must be an object, with the special style key inside. Which will require users to create a new array of objects. Initially, the way I designed the table was especially to be able to take the raw data from an API/DB and use it inside the table with only the keys we want to display, without manipulating the data (or manipulating while we're displaying with the cellFormatter).

It really depends on the use case, if there's one important use case for that, we might support it, otherwise it doesn't look necessary. Most of the time, a user should be able to achieve what he wants with a custom cellFormatter.

I believe your implementation is fine for now, maybe just check `style` is an object before adding it in the element?
And maybe rename it with `_mdlRowProps`, while adding support for custom className and style? Having a generic `props` would also allow to add onClick event and other things on the `tr`

**Comment by Augustin82** [comment]
OK, I'm on it. Thanks for the feedback and suggestions =)

**Comment by Augustin82** [comment]
Updated. What do you think, @tleunen ?

File: src/DataTable/Table.js [hunk 2]
```diff
@@ -21,6 +21,23 @@ const propTypes = {
     shadow: PropTypes.number
 };
 
+const isString = (s) => typeof(s) === 'string' || s instanceof String;
+
+const getCustomRowProps = (row) => {
```
Comment (tleunen) [inline src/DataTable/Table.js]: hmm.. Why do we really want to check the received data?
There's a high chance the user know what to do when he uses `mdlRowProps`... I mean.. I hope he knows what he's doing ;p

File: src/DataTable/Table.js [hunk 3]
```diff
@@ -64,11 +81,20 @@ class Table extends React.Component {
...
-                    )}
+                    {realRows.map((row, idx) => {
+                        const customRowProps = getCustomRowProps(row);
+
+                        return (
+                            <tr
+                                key={row[rowKeyColumn] || row.key || idx}
+                                style={customRowProps.style}
+                                className={customRowProps.className}
+                                onClick={customRowProps.onClick}
```
Comment (tleunen) [inline src/DataTable/Table.js]: I believe we can just pass everything without checks... Something like: `{ ...(row.mdlRowProps || {}) }`

Comment (tleunen) [inline src/DataTable/Table.js]: Also, what about the old `row.className`?

File: src/DataTable/Table.js [hunk 4]
```diff
@@ -64,11 +81,20 @@ class Table extends React.Component {
...
-                        </tr>
-                    )}
+                    {realRows.map((row, idx) => {
+                        const customRowProps = getCustomRowProps(row);
+
+                        return (
+                            <tr
+                                key={row[rowKeyColumn] || row.key || idx}
+                                style={customRowProps.style}
+                                className={customRowProps.className}
```
Comment (tleunen) [inline src/DataTable/Table.js]: Also, what about the old row.className?

File: src/DataTable/Table.js (hunk 3 above)
Comment (Augustin82) [inline src/DataTable/Table.js]: I wanted to make sure that we wouldn't be causing weird out-of-our-control React errors by allowing the wrong types through to React props.

If you think this is needlessly complicated, I'll remove. Thanks!

File: src/DataTable/Table.js (hunk 4 above)
Comment (Augustin82) [inline src/DataTable/Table.js]: I assumed we were replacing it. I'll merge instead.

File: src/DataTable/Table.js (hunk 3 above)
Comment (tleunen) [inline src/DataTable/Table.js]: `mdlRowProps` would be the responsibility of the user

File: src/DataTable/Table.js (hunk 4 above)
Comment (Augustin82) [inline src/DataTable/Table.js]: You're right, that's a pretty big miss. Sorry about that.

Which one would you favor? Adding to mdlRowProps (creating it in the process) or merging two separate objects?

I'd tend to favor keeping the current behaviour, only adding more classNames to it via mdlRowProps, in case people like me were already overloading row.className. What do you think?

Comment (tleunen) [inline src/DataTable/Table.js]: For now, merging them would be best. We'll rethink this later (hopefully for v2)

**Comment by Augustin82** [comment]
Updated as per your comments. Thoughts?

**Comment by Augustin82** [comment]
@tleunen sorry to ping you, any update on this?

File: src/DataTable/Table.js [hunk 5]
```diff
@@ -64,11 +64,19 @@ class Table extends React.Component {
...
-                        </tr>
-                    )}
+                    {realRows.map((row, idx) => {
+                        const { style, className: mdlRowPropsClassName, onClick } = row.mdlRowProps || {};
+
+                        return (
+                            <tr
+                                key={row[rowKeyColumn] || row.key || idx}
+                                className={classNames(row.className, mdlRowPropsClassName)}
+                                {...{ style, onClick }}
```
Comment (tleunen) [inline src/DataTable/Table.js]: Why not passing all the other `mdlRowProps`, not just style and onClick?

**Comment by tleunen** [comment]
Sorry I was away this weekend and came back yesterday, will make sure we merge this asap ;)

**Comment by Augustin82** [comment]
Is there any problem with it?

**Comment by Augustin82** [comment]
@tleunen any update on this, please?

**Comment by tleunen** [comment]
I thought I asked why doing `{...{ style, onClick }}` instead of just passing all props `{ ...mdlRowProps }` because if the user wants another event on it, like `onDoubleClick` or `onMouseOver`, we'll also need to add them. We can't really maintain a whitelist imo.

**Comment by Augustin82** [comment]
Apologies, I didn't understand it that way.

I have pushed an update, is that what you mean?

Thanks!

**Comment by tleunen** [comment]
Thanks @Augustin82 and sorry again for the long long delay

**Comment by Augustin82** [comment]
No need to apologize. Thank you very much for taking the time to review it, suggest correction, review it again because I messed up, and merging it ;)
